static/js/publisher/metrics/graphs/activeDevices.js

```diff
@@ -3,418 +3,730 @@ import { format } from "d3-format";
+        .pop();
+      currentHoverKey = match;
+    }
 
-    const current = d.key;
-    const dateData = _data[i];
+    const dateData = this.data[i];
 
-    tooltip
-      .html(tooltipTemplate(dateData, current, colorScale))
+    this.tooltip
+      .html(this.tooltipTemplate(dateData, currentHoverKey))
       .style("top", `${mousePosition[1]}px`)
-      .style("left", `${mousePosition[0] + margin.left}px`)
+      .style("left", `${mousePosition[0] + this.margin.left}px`)
       .style("display", "block");
-  }
 
-  function cancelTooltip() {
-    tooltip.style("display", "none");
-  }
+    const tooltipBounding = this.tooltip
+      .select(".p-tooltip__message")
+      .node()
+      .getBoundingClientRect();
 
-  const categories = document.querySelector(`[data-js="annotations-hover"]`);
-  categories.addEventListener("mouseover", e => {
-    const annotationHover = e.target.closest(`[data-js="annotation-hover"]`);
-    if (annotationHover) {
-      const category = annotationHover.dataset.id;
-      g.selectAll(`#${category}`).style("visibility", "visible");
-    }
-  });
+    if (tooltipBounding.top < 0) {
+      this.tooltip
+        .select(".p-tooltip")
+        .classed("p-tooltip--top-center", false)
+        .classed("p-tooltip--btm-center", true);
 
-  categories.addEventListener("mouseout", e => {
-    const annotationHover = e.target.closest(`[data-js="annotation-hover"]`);
-    if (annotationHover) {
-      const category = annotationHover.dataset.id;
-      g.selectAll(`#${category}`).style("visibility", "hidden");
+      this.tooltip.style("top", `${mousePosition[1] + 30}px`);
     }
-  });
-
-  showGraph(holder);
-}
+  }
 
-function activeDevices(
-  holderSelector,
-  activeDevices,
-  type,
-  defaultTrack,
-  annotations
-) {
-  const holder = document.querySelector(holderSelector);
-  graphType = type;
-  metricsDefaultTrack = defaultTrack;
-
-  if (!holder) {
-    return;
+  cancelTooltip() {
+    this.tooltip.style("display", "none");
   }
 
-  drawGraph(holderSelector, holder, activeDevices, annotations);
+  show() {
+    this.holder.style.opacity = 1;
 
-  const resize = debounce(function() {
-    drawGraph(holderSelector, holder, activeDevices, annotations);
-  }, 100);
+    return this;
+  }
 
-  select(window).on("resize", resize);
+  enableTooltip() {
+    const dataLayer = select(".layer");
+    const dataLayerBox = dataLayer.node().getBBox();
+
+    dataLayer
+      .append("rect")
+      .style("fill", "none")
+      .style("pointer-events", "all")
+      .attr("class", "pointerLayer")
+      .attr("width", dataLayerBox.width)
+      .attr("height", dataLayerBox.height + this.margin.top)
+      .attr("x", dataLayerBox.x)
+      .attr("y", dataLayerBox.y - this.margin.top / 2)
+      .on("mousemove", this.mouseMove.bind(this))
+      .on("mouseout", this.cancelTooltip.bind(this));
+
+    this.showTooltips = true;
+
+    return this;
+  }
 }
 
-export { activeDevices as default };
+export { Graph };
```

Comment:
Does `Graph` have any "active devices specific" functionality? It seems to be used elsewhere, so maybe the file could be renamed?

Reply:
Or maybe the graph should be called ActiveDevicesGraph if that's specific implementation that shows this particular type of data.